Problems with Hokuyo 04LX

Hi,
I'm trying to use the URG-04LX sensor, with the new urg_node. When I try to connect I get:
[ERROR] [1481563459.541562213]: Unknown error connecting to Hokuyo

While, if I'm using the old hokuyo_node, the sensor just work correctly.
Also, if I try to connect using the urg_node after killing the hokuyo node, the new node works.

 I've tried with different parameters and disabling/enabling calibration

Also I've updated the sensor to the latest version (Ver3.4.03) without success.

Restarting the sensor doesn't work, as it enters in a loop of error states.

rosrun urg_node urg_node 
[ INFO] [1481562698.401506634]: Connected to serial device with ID: receive error.
[ INFO] [1481562698.528700165]: Streaming data.
[ WARN] [1481562700.164722195]: Could not grab single echo scan.
[ERROR] [1481562706.575695889]: Error count exceeded limit, reconnecting.
[ INFO] [1481562710.064727728]: Connected to serial device with ID: receive error.
[ERROR] [1481562710.165523797]: Tried to advertise a service that is already advertised in this node [/urg_node/set_parameters]
[ INFO] [1481562710.187120186]: Streaming data.
[ WARN] [1481562711.828255910]: Could not grab single echo scan.
[ERROR] [1481562718.239361077]: Error count exceeded limit, reconnecting.
[ INFO] [1481562721.728254595]: Connected to serial device with ID: receive error.
[ INFO] [1481562721.849091216]: Streaming data.
[ WARN] [1481562723.491826093]: Could not grab single echo scan.
[ERROR] [1481562729.903079945]: Error count exceeded limit, reconnecting.
[ INFO] [1481562733.392075814]: Connected to serial device with ID: receive error.
[ERROR] [1481562733.492562034]: Tried to advertise a service that is already advertised in this node [/urg_node/set_parameters]
[ INFO] [1481562733.516776507]: Streaming data.
[ WARN] [1481562735.155774107]: Could not grab single echo scan.
[ERROR] [1481562741.566833458]: Error count exceeded limit, reconnecting.

Update: using the verison in the pull request #22 i get a more informative error:

$ rosrun urg_node urg_node 
[ERROR] [1482329968.746021512]: Error connecting to Hokuyo: Could not initialize Hokuyo:
not connected.


$ rosrun urg_node getID /dev/ttyACM0
getID failed: Could not initialize Hokuyo:
not connected.

However the hokuyo is connected and in fact works with the old hokuyo_node:
$ rosrun hokuyo_node hokuyo_node 
[ INFO] [1482330147.593798394]: Connected to device with ID: H0707625
[ INFO] [1482330147.692804670]: Starting calibration. This will take up a few seconds.
[ INFO] [1482330150.783847053]: Calibration finished. Latency is: -0.0810
[ INFO] [1482330151.093119302]: Streaming data.

the hokuyo is on the default serial over usb port:
$ ls /dev/ttyACM*
/dev/ttyACM0

--- a/include/urg_node/urg_c_wrapper.h
+++ b/include/urg_node/urg_c_wrapper.h
@@ -179,6 +179,12 @@ private:
   ros::Duration getTimeStampOffset(size_t num_measurements);
 
   /**
+   * @brief Set the Hokuyo URG-04LX from SCIP 1.1 mode to SCIP 2.0 mode.
+   * @returns True if successful and false if not.
+   */
+  bool setToSCIP2();
+
+  /**
    * @brief calculate the crc of a given set of bytes.
    * @param bytes The bytes array to be processed.
    * @param size The size of the bytes array.

--- a/src/urg_c_wrapper.cpp
+++ b/src/urg_c_wrapper.cpp
@@ -99,13 +99,22 @@ void URGCWrapper::initialize(bool& using_intensity, bool& using_multiecho)
   // urg_max_data_size can return a negative, error code value. Resizing based on this value will fail.
   if (urg_data_size < 0)
   {
-    urg_.last_errno = urg_data_size;
-    std::stringstream ss;
-    ss << "Could not initialize Hokuyo:\n";
-    ss << urg_error(&urg_);
-    stop();
-    urg_close(&urg_);
-    throw std::runtime_error(ss.str());
+    // This error can be caused by a URG-04LX in SCIP 1.1 mode, so we try to set SCIP 2.0 mode.
+    if (setToSCIP2() && urg_max_data_size(&urg_) >= 0)
+    {
+      // If setting SCIP 2.0 was successful, we set urg_data_size to the correct value.
+      urg_data_size = urg_max_data_size(&urg_);
+    }
+    else
+    {
+      urg_.last_errno = urg_data_size;
+      std::stringstream ss;
+      ss << "Could not initialize Hokuyo:\n";
+      ss << urg_error(&urg_);
+      stop();
+      urg_close(&urg_);
+      throw std::runtime_error(ss.str());
+    }
   }
   // Ocassionally urg_max_data_size returns a string pointer, make sure we don't allocate too much space,
   // the current known max is 1440 steps
@@ -519,6 +528,29 @@ bool URGCWrapper::getDL00Status(UrgDetectionReport& report)
   return true;
 }
 
+bool URGCWrapper::setToSCIP2()
+{
+  if (urg_.connection.type == URG_ETHERNET)
+    return false;
+
+  char buffer[sizeof("SCIP2.0\n")];
+  int n;
+
+  do {
+  n = serial_readline(&(urg_.connection.serial), buffer, sizeof(buffer), 1000);
+  } while (n >= 0);
+
+  serial_write(&(urg_.connection.serial), "SCIP2.0\n", sizeof(buffer));
+  n = serial_readline(&(urg_.connection.serial), buffer, sizeof(buffer), 1000);
+
+  // Check if switching was successful.
+  if (n > 0 && strcmp(buffer, "SCIP2.0") == 0 && urg_open(&urg_, URG_SERIAL, serial_port_.c_str(), (long)serial_baud_) >= 0)
+  {
+    ROS_DEBUG_STREAM("Set sensor to SCIP 2.0.");
+    return true;
+  }
+  return false;
+}
 
 uint16_t URGCWrapper::checkCRC(const char* bytes, const uint32_t size)
 {

